### ISSUE TITLE
Incorrect noImplicitAny return type warning when function expression is contextually typed by a union with a type parameter and returns a literal

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#31114]
**TypeScript Version:**  master through 2.0

**Search Terms:** noImplicitAny literal return

**Code**

```ts
declare function foo<T>(value: T | number): void;
foo((a: any) => "")
```

Criteria:
 * `value` is a union type
 * Return type of function expression is a literal type (array, boolean, string, number)

**Expected behavior:** No error

**Actual behavior:** Function expression has implicit 'any' return type because it is referenced in its return expression

**Playground Link:** http://www.typescriptlang.org/play/#src=%2F%2F%20%40noImplicitAny%3A%20true%0D%0Adeclare%20function%20foo%3CT%3E(value%3A%20T%20%7C%20number)%3A%20void%3B%0D%0Afoo((a%3A%20any)%20%3D%3E%20%22%22)%0D%0A%0D%0A

**Related Issues:** ?
